### ISSUE TITLE
NCCL error using DDP

### ISSUE THREAD

**Issue by zhangdan8962** [PyTorchLightning/pytorch-lightning#6527]
## 🐛 Bug

Getting this error when attempting to use ddp with VAE:
```
Traceback (most recent call last):
  File "/home/dzhang4/VAE/main.py", line 32, in <module>
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    results = self.accelerator_backend.train()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 303, in ddp_train
    self.barrier('ddp_setup')
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 186, in barrier
    torch_distrib.barrier(group=self.ddp_plugin.data_parallel_group)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    work = _default_pg.barrier()
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1607370116979/work/torch/lib/c10d/ProcessGroupNCCL.cpp:784, unhandled system error, NCCL version 2.7.8
```

### To Reproduce

```python
from pl_bolts.models.autoencoders import VAE
from pl_bolts.datamodules import ImagenetDataModule
import pytorch_lightning as pl
import torchvision.datasets as datasets
import torch
import torchvision.transforms as transforms
import os

model = VAE(224,'resnet50',enc_out_dim=2048,latent_dim=256)
traindir = os.path.join('/home/dzhang4/Imagenet', 'train')
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                 std=[0.229, 0.224, 0.225])
train_dataset = datasets.ImageFolder(
    traindir,
    transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            normalize,
        ]))


train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)


trainer = pl.Trainer(gpus=[2,3,4,5,6,7,8,9],profiler=True,accelerator='ddp')
trainer.fit(model, train_dataloader=train_loader)
```

### Expected behavior

Works on multi-GPUs

### Environment

* CUDA:
	- GPU:
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
		- Tesla V100-SXM3-32GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1
	- pytorch-lightning: 1.1.8
	- tqdm:              4.46.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.10
	- version:           #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019

**Comment by awaelchli** [comment]
I had the same problem with DDP before when selecting device indices. Updating to torch 1.8 solved the problem for me. Let me know if that helps.

**Comment by zhangdan8962** [comment]
I am not able to update to torch 1.8 since the cuda version is 10.1 on my server.

I was able to run with DP but it took  5 hours for one epoch. I am wondering if such cost sounds reasonable or not?

**Comment by awaelchli** [comment]
Try  `NCCL_IB_DISABLE=1`

See this issue #4420

**Comment by zhangdan8962** [comment]
I tried the method mentioned at https://github.com/PyTorchLightning/pytorch-lightning/issues/4420#issuecomment-788241369 and got new errors.

```
Traceback (most recent call last):
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
  File "/home/dzhang4/VAE/main.py", line 32, in <module>
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    results = self.accelerator_backend.train()
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    results = self.accelerator_backend.train()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
    results = self.accelerator_backend.train()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
    results = self.accelerator_backend.train()
    results = self.ddp_train(process_idx=self.task_idx, model=model)
    results = self.accelerator_backend.train()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
    self.trainer.is_slurm_managing_tasks
    results = self.accelerator_backend.train()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
    self.trainer.is_slurm_managing_tasks
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
    is_slurm_managing_tasks,
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
    results = self.ddp_train(process_idx=self.task_idx, model=model)
    self.trainer.is_slurm_managing_tasks
    trainer.fit(model, train_dataloader=train_loader)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
    is_slurm_managing_tasks,
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 510, in fit
    self.trainer.is_slurm_managing_tasks
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
    is_slurm_managing_tasks,
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    is_slurm_managing_tasks,
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    self.trainer.is_slurm_managing_tasks
    barrier()
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    results = self.accelerator_backend.train()
    self.trainer.is_slurm_managing_tasks
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 158, in train
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    results = self.ddp_train(process_idx=self.task_idx, model=model)
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/ddp_accelerator.py", line 261, in ddp_train
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    is_slurm_managing_tasks,
    is_slurm_managing_tasks,
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
    self.trainer.is_slurm_managing_tasks
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 163, in init_ddp_connection
    torch_backend, rank=global_rank, world_size=world_size
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    is_slurm_managing_tasks,
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/pytorch_lightning/plugins/ddp_plugin.py", line 94, in init_ddp_connection
    torch_backend, rank=global_rank, world_size=world_size
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer
    work = _default_pg.barrier()
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
RuntimeError: Connection reset by peer
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    barrier()
  File "/home/dzhang4/miniconda3/envs/coco/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer
    work = _default_pg.barrier()
RuntimeError: Connection reset by peer

**Comment by srib** [comment]
@zhangdan8962 What do you get when you use `PL_TORCH_DISTRIBUTED_BACKEND=gloo`? Could you please try and report?

**Comment by zhangdan8962** [comment]
Yes. It is working.

What are the downsides of gloo comparing nccl?

**Comment by srib** [comment]
> What are the downsides of gloo comparing nccl?

Thanks! Please see this [link](https://pytorch.org/docs/stable/distributed.html) comparing `gloo` and `nccl` and explaining in detail about those.

@awaelchli was kind enough to fix several issues in #6941. PL 1.2.8 might fix some of the issues that you are facing.

**Comment by awaelchli** [comment]
Thanks @srib for the help here.
@zhangdan8962 I understand that your issue is resolved?
